pkg/operator/ceph/cluster/osd/osd_pvc_test.go

```diff
@@ -0,0 +1,692 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package osd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/coreos/pkg/capnslog"
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	rookv1 "github.com/rook/rook/pkg/apis/rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+	cephver "github.com/rook/rook/pkg/operator/ceph/version"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	"github.com/rook/rook/pkg/operator/test"
+	exectest "github.com/rook/rook/pkg/util/exec/test"
+	"github.com/stretchr/testify/assert"
+	"github.com/tevino/abool"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/version"
+	"k8s.io/apimachinery/pkg/watch"
+	fakediscovery "k8s.io/client-go/discovery/fake"
+	"k8s.io/client-go/kubernetes/fake"
+	k8stesting "k8s.io/client-go/testing"
+)
+
+// make infof function for helping debugging by identifying logs output by the test itself
+func infof(t *testing.T, format string, args ...interface{}) {
+	logger.Infof(t.Name()+": "+format, args...)
+}
+
+// pick a node from a list of nodes in a round-robin fashion
+// use a pointer to slice so that the slice can be modified in-place
+func pickNode(nodes *[]string) string {
+	node := (*nodes)[0]
+	*nodes = append((*nodes)[1:], node)
+	return node
+}
+
+// The definition for this test is a wrapper for the test function that adds a timeout
+func TestOSDsOnPVC(t *testing.T) {
+	oldLogger := *logger
+	defer func() { logger = &oldLogger }() // reset logger to default after this test
+	logger.SetLevel(capnslog.TRACE)        // want more log info for this test if it fails
+
+	oldOpportunisticDuration := osdOpportunisticUpdateDuration
+	defer func() { osdOpportunisticUpdateDuration = oldOpportunisticDuration }()
+	// lower the opportunistic update check duration for unit tests to speed them up
+	osdOpportunisticUpdateDuration = 1 * time.Millisecond
+
+	// This test runs in less than 250 milliseconds on a 6-core CPU w/ 16GB RAM
+	// GitHub CI runner seems to be a lot slower (~10x longer)
+	done := make(chan bool)
+	timeout := time.After(25 * 250 * time.Millisecond)
```

Comment:
We might want to increase this again as the test fails with `2021-02-10T17:30:16.4923675Z     osd_pvc_test.go:93: Test timed out. This is a test failure.`.

Reply:
I've increased the timeout quite a bit but still see this timing out at the same place. I wonder if there is a race somewhere in here that is exacerbated in slower environments like CI. Makes it harder to debug, but I hope to find it.

Reply:
Hmm... I see the race. It's related to some of the changes I made related to creating pods for jobs in the reusable "complex" clientset... I'll have to either revert that or figure out a way to make sure the tests do the pod stuff ASAP.

Reply:
I fixed the race and cleaned up a few more things while I was figuring out how to refactor the job reactor.